### PR TITLE
(SERVER-2541) compile endpoint - clojure work

### PR DESCRIPTION
**Pull request by justinstoller**
I think this, minus cleaning up commit messages is ready to do. Hopefully this works with what you're working on @Magisus.

### CODE REVIEW

File: src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj [hunk 1]
```diff
@@ -81,6 +81,10 @@
    [this jruby-instance request-options]
    (.compileCatalog jruby-instance request-options))
 
+  (compile
```
Comment (justinstoller) [inline src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj]: this conflicts with a clojure var, I think I'll rename it to `compile-ast`, you're working on the Ruby side, right @Magisus ? Do you have thoughts on the protocol name? Should I implement the java interface as part of this work too?

Comment (justinstoller) [reply src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj]: renamed to `compile-ast` and implemented interface per convo in slack.

File: src/clj/puppetlabs/services/master/master_core.clj [hunk 2]
```diff
@@ -852,13 +897,19 @@
   (let [v4-catalog-handler (v4-catalog-handler
                              jruby-service
                              wrap-with-jruby-queue-limit
-                             current-code-id-fn)]
+                             current-code-id-fn)
+        compile-handler' (compile-handler
```
Comment (Magisus) [inline src/clj/puppetlabs/services/master/master_core.clj]: So I notice that we put the transports endpoint under v3 and this one under v4. Is there any policy we want to follow with choosing one of those going forward?

Comment (justinstoller) [reply src/clj/puppetlabs/services/master/master_core.clj]: Sorry, missed this earlier.

I put them under versions that were similar to other work that we were doing, but that's not really a reason for _why_ they should be there. I could see this being a `v3` endpoint.

Comment (Magisus) [reply src/clj/puppetlabs/services/master/master_core.clj]: We're already not doing regular API versioning (where the whole API is supposed to move to the new version when a change happens like the new `catalog` endpoint), so I guess it doesn't matter that much to anyone except us.

I think that given we just put the other net-new endpoint under v3, maybe this should go there too. It seems a little easier to remember that you only have to use `v4` for the one that has another meaning in `v3`.

Comment (justinstoller) [reply src/clj/puppetlabs/services/master/master_core.clj]: moved this to the v3 api

**Comment by justinstoller** [comment]
Did you still want review from @pcarlisle or @tvpartytonight, @Magisus ?

**Comment by Magisus** [comment]
No that's fine. But I just realized as I was working on my other PR, can you move the docs under the v3 directory as part of this?

File: src/clj/puppetlabs/services/master/master_core.clj [hunk 3]
```diff
@@ -712,34 +712,50 @@
...
+   (schema/required-key "code_ast") schema/Str
+   (schema/required-key "trusted_facts") {(schema/required-key "values") {schema/Str schema/Any}}
+   (schema/required-key "facts") {(schema/required-key "values") {schema/Str schema/Any}}
+   (schema/required-key "variables") {(schema/required-key "values") {schema/Str schema/Any}}
+   (schema/optional-key "job_id") schema/Str
+   (schema/optional-key "transaction_uuid") schema/Str
+   (schema/optional-key "options") {(schema/optional-key "capture_logs") schema/Bool
+                                    (schema/optional-key "log_level") (schema/enum "debug" "info" "warn")}})
+
+(defn valid-body
```
Comment (pcarlisle) [inline src/clj/puppetlabs/services/master/master_core.clj]: more clear if named validate-body, makes the name obviously parallel to schema/validate

Comment (pcarlisle) [reply src/clj/puppetlabs/services/master/master_core.clj]: oh i see you want the difference of returning the body after validating.... this is still a little weird but i'm not as sure

Comment (justinstoller) [reply src/clj/puppetlabs/services/master/master_core.clj]: `validated-body`?

**Comment by justinstoller** [comment]
updated

**Comment by Magisus** [comment]
Uh oh, conflict.

**Comment by justinstoller** [comment]
rebased, I think this is ready for merge

**Pull request merged**